generator/graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/generator/federation/execution/EntitiesDataFetcher.kt

```diff
@@ -0,0 +1,107 @@
+
+private const val TYPENAME_FIELD = "__typename"
+private const val REPRESENTATIONS = "representations"
+
+/**
+ * Federated _entities field data fetcher.
+ */
+open class EntitiesDataFetcher(
+    resolvers: List<TypeResolver>
+) : DataFetcher<CompletableFuture<DataFetcherResult<List<Any?>>>> {
+
+    constructor(vararg resolvers: TypeResolver) : this(resolvers.toList())
+    /**
+     * Pre-compute resolvers by typename so, we don't have to search on every request
+     */
+    private val resolversByType: Map<String, TypeResolver> = resolvers.associateBy(TypeResolver::typeName)
+
+    /**
+     * Resolves entities based on the passed in representations argument. Entities are resolved in the same order
+     * they are specified in the list of representations. If target representation cannot be resolved, NULL will
+     * be returned instead.
+     *
+     * Representations are grouped by the underlying typename and each batch is resolved asynchronously before merging
+     * the results back into a single list that preserves the original order.
+     *
+     * @return list of resolved nullable entities
+     */
+    override fun get(env: DataFetchingEnvironment): CompletableFuture<DataFetcherResult<List<Any?>>> {
+        val representations: List<Map<String, Any>> = env.getArgument(REPRESENTATIONS)
+
+        val representationsWithoutResolver = mutableListOf<IndexedValue<Map<String, Any>>>()
+        val entitiesWithPromiseResolver = mutableListOf<ResolvableEntity<FederatedTypePromiseResolver<*>>>()
+        val entitiesWithSuspendResolver = mutableListOf<ResolvableEntity<FederatedTypeResolver<*>>>()
+
+        representations.withIndex()
+            .groupBy { (_, representation) -> representation[TYPENAME_FIELD].toString() }
+            .forEach { (typeName, indexedRequests) ->
+                when (val resolver = resolversByType[typeName]) {
+                    is FederatedTypePromiseResolver<*> -> {
+                        entitiesWithPromiseResolver += ResolvableEntity(typeName, indexedRequests, resolver)
+                    }
+                    is FederatedTypeResolver<*> -> {
+                        entitiesWithSuspendResolver += ResolvableEntity(typeName, indexedRequests, resolver)
+                    }
+                    null -> {
+                        representationsWithoutResolver += indexedRequests
+                    }
+                }
+            }
+
+        val noResolverErrors: CompletableFuture<List<Map<Int, Any?>>> = CompletableFuture.completedFuture(
+            listOf(
+                representationsWithoutResolver.map(IndexedValue<Map<String, Any>>::index).zip(
+                    representationsWithoutResolver.map(IndexedValue<Map<String, Any>>::value).map { representation ->
+                        InvalidFederatedRequest("Unable to resolve federated type, representation=$representation")
+                    }
+                ).toMap()
+            )
+        )
```

Comment:
I think you can solve this with [associateWith](https://kotlinlang.org/docs/collection-transformations.html#associate)??

If not maybe a variable or two in this block would make it seem a little less magic or deep kotlin knowledge

Reply:
```kotlin
representationsWithoutResolver.associateBy(IndexedValue<Map<String, Any>>::index) { (_, representation) ->
                    InvalidFederatedRequest("Unable to resolve federated type, representation=$representation")
                }
```